afl-fuzz.c

```diff
@@ -1306,6 +1312,63 @@ static void update_bitmap_score(struct queue_entry* q) {
 
 }
 
+double rand_double() 
+{
+  // return random value in interval [0.0,1.0)
+  double rnd = (double)UR(RAND_MAX);
+  double max = (double)RAND_MAX;
+  return rnd / max;
+}
+
+// find the first element in array that is greater or equal target
+int first_greater_or_equal_element(double arr[], double target, int end)
+{
+    int start = 0;
+    int ans = -1;
+    while (start <= end)
+    {
+        int mid = (start + end) / 2;
+        if (arr[mid] < target) {
+            start = mid + 1;
+        } else {
+            ans = mid;
+            end = mid - 1;
+        }
+    }
+    return ans;
+}
+
+
+static void mark_selected_inputs() {
+  double cumulative_sum[queued_paths];
```

Comment:
I think it might be more clean to allocate `queue_list` here instead of globally. What do you think?

Then the size is always `queued_paths`.

Reply:
Something like this:
```
double cumulative_sum[queued_paths];
queue_entry* queue_entry[queued_paths];
int idx = 0;
q = queue;
while (q) {
  double w = 1.0;
  q->is_selected = 0; // reset flag from previous cycle
  if (q->favored) {
    w *= 20.0;
  } else if (!q->was_fuzzed) { 
    w *= 5.0;
  }
  total_weight += w;
  cumulative_sum[i] = total_weight;
  queue_entry[idx] = q;
  q = q->next;
  idx++;
}
```

Reply:
Ah yes, this is way more clean than my implementation. Thanks!